rust/ql/lib/codeql/rust/internal/typeinference/FunctionType.qll

```diff
@@ -0,0 +1,434 @@
+      or
+      parent = i.(TraitItemNode).resolveBound(parentMention)
+    )
+  }
+
+/**
+ * The type of an associated function at a given position, when viewed as a member
+ * of a given trait or `impl` block.
+ *
+ * Example:
+ *
+ * ```rust
+ * trait T1 {
+ *   fn m1(self);              // self1
+ *
+ *   fn m2(self) { ... }       // self2
+ * }
+ *
+ * trait T2 : T1 {
+ *   fn m3(self);              // self3
+ * }
+ *
+ * impl T2 for X {
+ *   fn m1(self) { ... }       // self4
+ *
+ *   fn m3(self) { ... }       // self5
+ * }
+ * ```
+ *
+ * param   | `impl` or trait | type
+ * ------- | --------------- | ----
+ * `self1` | `trait T1`      | `T1`
+ * `self1` | `trait T2`      | `T2`
+ * `self2` | `trait T1`      | `T1`
+ * `self2` | `trait T2`      | `T2`
+ * `self2` | `impl T2 for X` | `X`
+ * `self3` | `trait T2`      | `T2`
+ * `self4` | `impl T2 for X` | `X`
+ * `self5` | `impl T2 for X` | `X`
+ */
+class AssocFunctionType extends TAssocFunctionType {
+  private predicate isFunctionType(Function f, FunctionTypePosition pos, ImplOrTraitItemNode i) {
+    this = MkAssocFunctionType(f, pos, i)
+  }
+
+  private predicate isInheritedFunctionType(
+    Function f, FunctionTypePosition pos, TypeMention parentMention, ImplOrTraitItemNode parent,
+    ImplOrTraitItemNode i
+  ) {
+    this = MkInheritedAssocFunctionType(f, pos, parentMention, parent, i)
+  }
+
+  /**
+   * Holds if this function type applies to the function `f` at position `pos`,
+   * when viewed as a member of the `impl` or trait item `i`.
+   */
+  predicate appliesTo(Function f, FunctionTypePosition pos, ImplOrTraitItemNode i) {
+    this.isFunctionType(f, pos, i)
+    or
+    this.isInheritedFunctionType(f, pos, _, _, i)
+  }
+
+  /** Gets the type at the given path. */
+  pragma[nomagic]
+  Type getDeclaredTypeAt(TypePath path) {
+    exists(Function f, FunctionTypePosition pos |
+      this.isFunctionType(f, pos, _) and
+      result = pos.getTypeMention(f).resolveTypeAt(path)
+    )
+    or
+    exists(
+      Function f, FunctionTypePosition pos, TypeMention parentMention, ImplOrTraitItemNode parent,
+      AssocFunctionType parentType, ImplOrTraitItemNode i
+    |
+      this.isInheritedFunctionType(f, pos, parentMention, parent, i) and
+      parentType.appliesTo(f, pos, parent)
+    |
+      result = parentType.getDeclaredTypeAt(path) and
+      not result instanceof TypeParameter
+      or
+      exists(TypePath prefix, TypePath suffix | path = prefix.append(suffix) |
+        parentType.hasSelfTypeParameterAt(prefix) and
+        result = resolveImplOrTraitType(i, suffix)
+        or
+        exists(TypeParameter tp |
+          parentType.hasTypeParameterAt(prefix, tp) and
+          result = parentMention.resolveTypeAt(TypePath::singleton(tp).appendInverse(suffix))
+        )
+      )
+    )
+  }
+
+  pragma[nomagic]
+  private predicate hasTypeParameterAt(TypePath path, TypeParameter tp) {
+    this.getDeclaredTypeAt(path) = tp
```

Comment:
This would be natural as a predicate with result:
```suggestion
  private TypeParameter getTypeParameterAt(TypePath path) {
    result = this.getDeclaredTypeAt(path)
```